pkgs/applications/misc/syncthingtray/default.nix

```diff
@@ -0,0 +1,50 @@
+{ mkDerivation, stdenv, lib, fetchFromGitHub
+, qtbase, extra-cmake-modules, cpp-utilities, qtutilities
+, cmake, kio, plasma-framework, qttools
+, webviewProvider ? null
```

Comment:
Why does this default to `null`? What if it isn't given?

Reply:
From my poor experience in packaging and reading other packages' arguments, this is what I figure understood: If this argument is `null`, when added to the `buildInputs` list, it's as if it wasn't added at all and so the compilation will consider this optional requirement as missing.

Reply:
Well kind of. When `buildInputs` is converted to a bash `PATH`, `null` values are ignored. But you shouldn't rely on that. For example when writing a nix function that does something with `buildInputs`, it would be ugly to have to deal with `null`s. Its also a bit implicit.

If its really optional, there should instead be an `xyzSupport` flag.

Reply:
I understand but this case is a bit more complicated then a ON/OFF switch - there are 2 packages that can provide webview and JS support.

Is there a way to check if these arguments are `null` and only if then, add them to `buildInputs`?

Reply:
Yes, you can check if they are `null`, but that's not a good way to do it. `callPackage` could "accidentally" pass a non-null value and enable a feature the user doesn't want enabled.

You could for example have

```
{ webviewPackage ? packageA
, enableWebview ? true
, ...
}:
```

Or alternatively default to `null` but add
```
assert enableWebview -> webviewPackage != null
```

If the differences between the webview implementations aren't important, it may also be reasonable to just choose one and not offer the choice.

Reply:
> If the differences between the webview implementations aren't important, it may also be reasonable to just choose one and not offer the choice.

IMO the difference is rather significant, both webview packages are pretty big so a user may wish to choose the implementation that he uses with other packages as well.

I'm not sure I understood your suggestion. Could you please give a fuller example / diff? Thanks.

Reply:
Since the author chimed in (thanks @Martchus) I think we should stick to the well-supported variants, i.e. `qtdeclarative` and `webengine`. So in that case you'd just have:

```
{ qtdeclarative
, qtwebengine
, webViewSupport ? true # or whatever you deem good as a default
, jsSupport ? true
}:
```

And then add `qtdeclarative` and `qtwebengine` to `buildInputs` conditionally with `lib.optionals`.